README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Build Status](https://travis-ci.org/hrbrmstr/cymruservices.svg)](https://travis-ci.org/hrbrmstr/cymruservices) [![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/0.1.0/active.svg)](http://www.repostatus.org/#active) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/cymruservices)](http://cran.r-project.org/package=cymruservices) ![downloads](http://cranlogs.r-pkg.org/badges/grand-total/cymruservices)

![img](dragonr.jpg)

cymruservices is an R package that provides interfaces to various [Team Cymru Services](http://www.team-cymru.org/services.html) including:

-   [The Bogon Reference](http://www.team-cymru.org/bogon-reference.html)
-   [The IP to ASN Mapping Project](http://www.team-cymru.org/IP-ASN-mapping.html)
-   [The Malware Hash Registry](http://www.team-cymru.org/MHR.html)

The following functions are implemented:

-   `bulk_origin`: Retrieves BGP Origin ASN info for a list of IPv4 addresses
-   `bulk_origin_asn`: Retrieves BGP Origin ASN info for a list of ASN ids
-   `bulk_peer`: Retrieves BGP Peer ASN info for a list of IPv4 addresses
-   `ipv4_bogons`: Retrieve list of IPv4 "full bogons" from Team Cymru webservice
-   `ipv6_bogons`: Retrieve list of IPv6 "full bogons" from Team Cymru webservice
-   `malware_hash`: Retrieves malware hash metadata from the Malware Hash Registry
-   `flush`: flush one or more API caches

### News

-   Version 0.3.0 uses `memoise` to cache API responses. See `flush()` for details. functions that used to return `data.frame`s now return `tbl_df`s
-   Version 0.2.0 `timeout` parameter added to `whois`-based functions
-   Version 0.1.0 is on CRAN
-   Version 0.1.0.9999 released

### Installation

``` r
devtools::install_github("hrbrmstr/cymruservices")
# OR
install.packages("cymruservices")
```

### Usage

``` r
library(cymruservices)
#> 
#> Attaching package: 'cymruservices'
#> The following object is masked from 'package:base':
#> 
#>     flush

# current verison
packageVersion("cymruservices")
#> [1] '0.3.0'
```

### Test Results

``` r
library(cymruservices)
library(testthat)

date()
#> [1] "Sun Mar  6 09:35:09 2016"

# only using `force=TRUE` to ensure output for the example
# see the help for each function to see why this is a bad
# idea to run force all the time
head(ipv4_bogons(force=TRUE))
#> [1] "0.0.0.0/8"     "2.56.0.0/14"   "5.8.248.0/21"  "5.45.32.0/20"  "5.133.64.0/18" "5.180.0.0/14"

head(ipv6_bogons(force=TRUE))
#> [1] "::/8"     "100::/8"  "200::/7"  "400::/6"  "800::/5"  "1000::/4"

bulk_origin(c("68.22.187.5", "207.229.165.18", "198.6.1.65"))
#> Source: local data frame [3 x 7]
#> 
#>      as             ip       bgp_prefix    cc registry  allocated
#>   (dbl)          (chr)            (chr) (chr)    (chr)      (chr)
#> 1 23028    68.22.187.5   68.22.187.0/24    US     arin 2002-03-15
#> 2  6079 207.229.165.18 207.229.128.0/18    US     arin           
#> 3   701     198.6.1.65     198.6.0.0/16    US     arin           
#> Variables not shown: as_name (chr)

bulk_peer(c("68.22.187.5", "207.229.165.18", "198.6.1.65"))
#> Source: local data frame [13 x 7]
#> 
#>    peer_as             ip       bgp_prefix    cc registry  allocated                                   peer_as_name
#>      (dbl)          (chr)            (chr) (chr)    (chr)      (chr)                                          (chr)
#> 1      174    68.22.187.5   68.22.187.0/24    US     arin 2002-03-15          COGENT-174 - Cogent Communications,US
#> 2     2381    68.22.187.5   68.22.187.0/24    US     arin 2002-03-15                       WISCNET1-AS - WiscNet,US
#> 3     1273 207.229.165.18 207.229.128.0/18    US     arin                    CW Cable and Wireless Worldwide plc,GB
#> 4     2381 207.229.165.18 207.229.128.0/18    US     arin                                  WISCNET1-AS - WiscNet,US
#> 5     2603 207.229.165.18 207.229.128.0/18    US     arin                                      NORDUNET NORDUnet,NO
#> 6     3257 207.229.165.18 207.229.128.0/18    US     arin                                 GTT-BACKBONE Tinet Spa,DE
#> 7     3356 207.229.165.18 207.229.128.0/18    US     arin                  LEVEL3 - Level 3 Communications, Inc.,US
#> 8      174     198.6.1.65     198.6.0.0/16    US     arin                     COGENT-174 - Cogent Communications,US
#> 9     1299     198.6.1.65     198.6.0.0/16    US     arin                                TELIANET TeliaSonera AB,SE
#> 10    2914     198.6.1.65     198.6.0.0/16    US     arin            NTT-COMMUNICATIONS-2914 - NTT America, Inc.,US
#> 11    3257     198.6.1.65     198.6.0.0/16    US     arin                                 GTT-BACKBONE Tinet Spa,DE
#> 12    3356     198.6.1.65     198.6.0.0/16    US     arin                  LEVEL3 - Level 3 Communications, Inc.,US
#> 13    3549     198.6.1.65     198.6.0.0/16    US     arin               LVLT-3549 - Level 3 Communications, Inc.,US

bulk_origin_asn(c(22822, 1273, 2381, 2603, 2914, 3257, 3356, 11164,
                  174, 286, 1299, 2914, 3257, 3356, 3549, 22822))
#> Source: local data frame [16 x 5]
#> 
#>       as    cc registry  allocated                                        as_name
#>    (dbl) (chr)    (chr)      (chr)                                          (chr)
#> 1  22822    US     arin 2001-11-28             LLNW - Limelight Networks, Inc.,US
#> 2   1273    EU  ripencc                    CW Cable and Wireless Worldwide plc,GB
#> 3   2381    US     arin                                  WISCNET1-AS - WiscNet,US
#> 4   2603    NO  ripencc                                      NORDUNET NORDUnet,NO
#> 5   2914    US     arin            NTT-COMMUNICATIONS-2914 - NTT America, Inc.,US
#> 6   3257    DE  ripencc                                 GTT-BACKBONE Tinet Spa,DE
#> 7   3356    US     arin 2000-03-10       LEVEL3 - Level 3 Communications, Inc.,US
#> 8  11164    US     arin 2014-07-11       INTERNET2-TRANSITRAIL-CPS - Internet2,US
#> 9    174    US     arin                     COGENT-174 - Cogent Communications,US
#> 10   286    NL  ripencc                                           KPN KPN B.V.,NL
#> 11  1299    EU  ripencc                                TELIANET TeliaSonera AB,SE
#> 12  2914    US     arin            NTT-COMMUNICATIONS-2914 - NTT America, Inc.,US
#> 13  3257    DE  ripencc                                 GTT-BACKBONE Tinet Spa,DE
#> 14  3356    US     arin 2000-03-10       LEVEL3 - Level 3 Communications, Inc.,US
#> 15  3549    US     arin 2000-03-21    LVLT-3549 - Level 3 Communications, Inc.,US
#> 16 22822    US     arin 2001-11-28             LLNW - Limelight Networks, Inc.,US

malware_hash(c("1250ac278944a0737707cf40a0fbecd4b5a17c9d",
              "7697561ccbbdd1661c25c86762117613",
              "cbed16069043a0bf3c92fff9a99cccdc",
              "e6dc4f4d5061299bc5e76f5cd8d16610",
              "e1112134b6dcc8bed54e0e34d8ac272795e73d74"))
#> Source: local data frame [5 x 3]
#> 
#>                                   sha1_md5 last_known_timestamp detection_pct
#>                                      (chr)               (time)         (dbl)
#> 1 1250ac278944a0737707cf40a0fbecd4b5a17c9d  2016-03-06 14:35:12            NA
#> 2         7697561ccbbdd1661c25c86762117613  2016-03-06 14:35:12            NA
#> 3         cbed16069043a0bf3c92fff9a99cccdc  2016-03-06 14:35:12            NA
#> 4         e6dc4f4d5061299bc5e76f5cd8d16610  2016-03-06 14:35:12            NA
#> 5 e1112134b6dcc8bed54e0e34d8ac272795e73d74  2016-03-06 14:35:12            NA

test_dir("tests/")
#> testthat results ========================================================================================================
#> OK: 6 SKIPPED: 0 FAILED: 0
```

### Code of Conduct

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS
Changes in cymruservices 0.5.0
==============================

- Fixed minor CRAN issues.
- Updated tests.

Changes in cymruservices 0.4.0
==============================

- bumped up timeout on tests
- switched all `http` URLs to `https` where possible.
- added `cymru_active()` to enable testing of connectibyt to Team Cymru services.

Changes in cymruservices 0.3.0
==============================

- implicit API response caching has been added via the memoise package.
- there is a new `flush()` function which flushes the new API caches
- functions that used to return `data.frame`s now return `tbl_df`s

Changes in cymruservices 0.2.0
==============================

- cymru service instability was causing CRAN daily testing errors so
  examples were wrapped in \dontrun{} even though the socketConnection()
  calls were wrapped in tryCatch blocks.
- added a `timeout` parameter to the `whois`-based services
- calls that used to error out and return `NA` now return `NA`-filled `data.frame`s

Changes in cymruservices 0.1.0
==============================

- CRAN release
- `bulk_origin`:	Retrieves BGP Origin ASN info for a list of IPv4 addresses
- `bulk_origin_asn`:	Retrieves BGP Origin ASN info for a list of ASN ids
- `bulk_peer`:	Retrieves BGP Peer ASN info for a list of IPv4 addresses
- `ipv4_bogons`:	Retrieve list of IPv4 "full bogons" from Team Cymru webservice
- `ipv6_bogons`:	Retrieve list of IPv6 "full bogons" from Team Cymru webservice
- `malware_hash`:	Retrieves malware hash metadata from the Malware Hash Registry


tests/testthat/test-cymruservices.R
context("bogons")

test_that("bogons work", {
  expect_that(ipv4_bogons(force=TRUE), is_a("character"))
  expect_that(ipv6_bogons(force=TRUE), is_a("character"))
})

context("bulk_origin")

test_that("bulk_origin works", {
  expect_that(bulk_origin(c("68.22.187.5", "207.229.165.18", "198.6.1.65"),
                          timeout=5),
              is_a("data.frame"))
})

context("bulk_peer")

test_that("bulk_peer works", {

  testthat::skip_on_cran()
  testthat::skip_on_travis()

  expect_that(suppressWarnings(
    bulk_peer(c("68.22.187.5", "207.229.165.18", "198.6.1.65"), timeout=5)),
              is_a("data.frame"))
})

context("bulk_origin_asn")

test_that("bulk_origin_asn works", {
  expect_that(bulk_origin_asn(c(22822, 1273, 2381, 2603, 2914, 3257, 3356, 11164,
                                174, 286, 1299, 2914, 3257, 3356, 3549, 22822),
                              timeout=5),
              is_a("data.frame"))
})

context("malware_hash")

test_that("malware_hash works", {
  expect_that(malware_hash(c("1250ac278944a0737707cf40a0fbecd4b5a17c9d",
                             "7697561ccbbdd1661c25c86762117613",
                             "cbed16069043a0bf3c92fff9a99cccdc",
                             "e6dc4f4d5061299bc5e76f5cd8d16610",
                             "e1112134b6dcc8bed54e0e34d8ac272795e73d74"),
                           timeout=5),
              is_a("data.frame"))
})


R/bogons.R
# documented, memoised versions of these are at the bottom

.ipv4_bogons <- function(force=FALSE, cached_bogons=NA) {

  BOGONS_URL <- "https://www.team-cymru.org/Services/Bogons/fullbogons-ipv4.txt"

  if (Sys.getenv("CYMRU_LAST_V4_BOGON") == "") {

    bogons <- safe_rl(BOGONS_URL)

    if (is.null(bogons$result)) {
      message("Could not reach team-cymru.org")
      return(NA_character_)
    }

    bogons <- bogons$result

    last_updated <- as.numeric(str_match_all(bogons[1],
                                             "updated ([[:digit:]]+) ")[[1]][,2])

    Sys.setenv(CYMRU_LAST_V4_BOGON=last_updated)

    return(tail(bogons, -1))

  } else {

    delta <- (as.numeric(Sys.time()) - as.numeric(Sys.getenv("CYMRU_LAST_V4_BOGON")))

    if ((delta > 14400) | ((delta < 14400) & force)) {

      bogons <- safe_rl(BOGONS_URL)

      if (is.null(bogons$result)) {
        message("Could not reach team-cymru.org")
        return(NA_character_)
      }

      bogons <- bogons$result

      last_updated <- as.numeric(str_match_all(bogons[1],
                                               "updated ([[:digit:]]+) ")[[1]][,2])

      Sys.setenv(CYMRU_LAST_V4_BOGON=last_updated)

      return(tail(bogons, -1))

    } else {

      message("It has not been 4 hours since the last refresh. Use 'force=TRUE' to force a refresh.")

      return(cached_bogons)

    }

  }

}

.ipv6_bogons <- function(force=FALSE, cached_bogons=NA) {

  FULL_BOGONS_URL <- "https://www.team-cymru.org/Services/Bogons/fullbogons-ipv6.txt"

  if (Sys.getenv("CYMRU_LAST_V6_BOGON") == "") {

    bogons <- safe_rl(FULL_BOGONS_URL)

    if (is.null(bogons$result)) {
      message("Could not reach team-cymru.org")
      return(NA_character_)
    }

    bogons <- bogons$result

    last_updated <- as.numeric(str_match_all(bogons[1],
                                             "updated ([[:digit:]]+) ")[[1]][,2])

    Sys.setenv(CYMRU_LAST_V6_BOGON=last_updated)

    return(tail(bogons, -1))

  } else {

    delta <- (as.numeric(Sys.time()) - as.numeric(Sys.getenv("CYMRU_LAST_V6_BOGON")))

    if ((delta > 14400) | ((delta < 14400) & force)) {

      bogons <- safe_rl(FULL_BOGONS_URL)

      if (is.null(bogons$result)) {
        message("Could not reach team-cymru.org")
        return(NA_character_)
      }

      bogons <- bogons$result

      last_updated <- as.numeric(str_match_all(bogons[1],
                                               "updated ([[:digit:]]+) ")[[1]][,2])

      Sys.setenv(CYMRU_LAST_V6_BOGON=last_updated)

      return(tail(bogons, -1))

    } else {

      message("It has not been 4 hours since the last refresh. Use 'force=TRUE' to force a refresh.")

      return(cached_bogons)

    }

  }
}

#' Retrieve list of IPv4 "full bogons" from Team Cymru webservice
#'
#' The traditional bogon prefixes (IPV4), plus prefixes that have been allocated to
#' RIRs but not yet assigned by those RIRs to ISPs, end-users, etc. Updated every four hours.
#'
#' Bogons are defined as Martians (private and reserved addresses defined by RFC 1918,
#' RFC 5735, and RFC 6598) and netblocks that have not been allocated to a regional
#' internet registry (RIR) by the Internet Assigned Numbers Authority.
#'
#' Fullbogons are a larger set which also includes IP space that has been allocated to
#' an RIR, but not assigned by that RIR to an actual ISP or other end-user. IANA maintains
#' a convenient IPv4 summary page listing allocated and reserved netblocks, and each RIR
#' maintains a list of all prefixes that they have assigned to end-users. Our bogon reference
#' pages include additional links and resources to assist those who wish to properly filter
#' bogon prefixes within their networks.
#'
#' @param force force a refresh even if the time-frame (4-hours) is not up
#' @param cached_bogons if you pass in the previous result of a call to \code{ipv4_bogoons}
#'        it will be returned if the refresh time constraint has not been met, otherwise
#'        \code{NA} will be returned.
#' @seealso \url{http://www.team-cymru.org/bogon-reference-http.html}
#' @export
#' @examples \dontrun{
#' v4_bogons <- ipv4_bogons()
#' v4_bogons <- ipv4_bogons(cached_bogons=v4_bogons)
#' }
ipv4_bogons <- memoise::memoise(.ipv4_bogons)

#' Retrieve list of IPv6 "full bogons" from Team Cymru webservice
#'
#' IPv6 "fullbogons", all IPv6 prefixes that have not been allocated to RIRs and
#' that have not been assigned by RIRs to ISPs, end-users, etc. Updated every four hours.
#'
#' Bogons are defined as Martians (private and reserved addresses defined by RFC 1918,
#' RFC 5735, and RFC 6598) and netblocks that have not been allocated to a regional
#' internet registry (RIR) by the Internet Assigned Numbers Authority.
#'
#' Fullbogons are a larger set which also includes IP space that has been allocated to
#' an RIR, but not assigned by that RIR to an actual ISP or other end-user. IANA maintains
#' a convenient IPv4 summary page listing allocated and reserved netblocks, and each RIR
#' maintains a list of all prefixes that they have assigned to end-users. Our bogon reference
#' pages include additional links and resources to assist those who wish to properly filter
#' bogon prefixes within their networks.
#'
#' @param force force a refresh even if the time-frame (4-hours) is not up
#' @param cached_bogons if you pass in the previous result of a call to \code{ipv6_bogoons}
#'        it will be returned if the refresh time constraint has not been met, otherwise
#'        \code{NA} will be returned.
#' @seealso \url{http://www.team-cymru.org/bogon-reference-http.html}
#' @export
#' @examples \dontrun{
#' v6_bogons <- ipv6_bogons()
#' v6_bogons <- ipv6_bogons(cached_bogons=v6_bogons)
#' }
ipv6_bogons <- memoise::memoise(.ipv6_bogons)


R/zflush.r
list(
  "origin" = bulk_origin,
  "peer" = bulk_peer,
  "asn" = bulk_origin_asn,
  "v4_bogons" = ipv4_bogons,
  "v6_bogons" = ipv6_bogons,
  "hash" = malware_hash
) -> valid_flushed

#' Flush cached results
#'
#' Within a given R session, it will be highly unlikely that API responses
#' to calls to Team Cymru services will change if the parameters have
#' not varied (i.e. you use the same vector of IP addresses again). To
#' respect the resources that have beeen freely provided, all the API
#' functions cache their results.\cr
#' \cr
#' It may be advantageous or necessary to invalidate one or more of these
#' caches. This function allows for the invalidation of one or more (or all)
#' caches.
#'
#' @param ... strings naming cached results to flush. Can be any of
#'        "\code{origin}", "\code{peer}", "\code{asn}", "\code{v4_bogons}",
#'        "\code{v6_bogons}" or "\code{hash}". If no parameters
#'        are specified all caches will be flushed.
#' @param quiet if \code{TRUE} no diagnostic or informative messages will be
#'        displayed. If \code{FALSE} warnings for unknown cache names and
#'        invalidation progress for valid caches will be displayed if
#'        the session is interactive.
#' @export
#' @note Invalid cache names will be ignored. If \code{quiet} is \code{FALSE}
#'       and \code{flush} was called from an interactive session invalid
#'       cache names will be noted.\cr\cr Also, you will still need to
#'       \code{force} the reloading of bogon lists if you are within the 4
#'       hour window even if you invalided the memoised cache.
#' @examples \dontrun{
#' flush("peer", "origin")
#' flush()
#' }
flush <- function(..., quiet=TRUE) {

  dots <- purrr::flatten_chr(list(...))

  can_flush <- names(valid_flushed)
  if (length(dots) == 0) { dots <- can_flush }

  confused <- setdiff(dots, can_flush)
  if (length(confused) > 0 & (!quiet) & interactive()) {
    warning(sprintf("Ignoring unrecognized caches: %s",
                    paste(sprintf("'%s'", confused), sep="", collapse=", ")))
  }

  ok <- intersect(can_flush, dots)
  if (length(ok) > 0) {
    purrr::walk(ok, function(x) {
      if ((!quiet) & interactive()) message(sprintf("Flushing '%s'...", x))
      memoise::forget(valid_flushed[[x]])
    })
  }

}



R/asn.R
# the documented, memoised versions of these are at the bottom

.bulk_origin <- function(ips, timeout=getOption("timeout")) {

  host <- "v4.whois.cymru.com"
  port <- 43

  # setup query
  cmd <- "begin\nverbose\n"
  ips_c <- paste(unlist(ips), collapse="\n")
  cmd <- sprintf("%s%s\nend\n", cmd, ips_c)

  # setup connection and post query
  con <- sock(host=host, port=port, blocking=TRUE, open="r+", timeout=timeout)
  if (is.null(con$result)) {
    message("Error opening connection to v4.whois.cymru.com")
    return(dplyr::tbl_df(data.frame(as=rep(NA, length(ips)),
                                    ip=rep(NA, length(ips)),
                                    bgp_prefix=rep(NA, length(ips)),
                                    cc=rep(NA, length(ips)),
                                    registry=rep(NA, length(ips)),
                                    allocated=rep(NA, length(ips)),
                                    as_name=rep(NA, length(ips)))))
  }

  con <- con$result
  cat(cmd, file=con)
  response <- readLines(con)
  close(con)

  if (length(response) == 0) {
    message("Error reading from connection to v4.whois.cymru.com")
    return(dplyr::tbl_df(data.frame(as=rep(NA, length(ips)),
                                    ip=rep(NA, length(ips)),
                                    bgp_prefix=rep(NA, length(ips)),
                                    cc=rep(NA, length(ips)),
                                    registry=rep(NA, length(ips)),
                                    allocated=rep(NA, length(ips)),
                                    as_name=rep(NA, length(ips)))))
  }

  # trim header, split fields and convert results
  response <- trim_df(read.csv(textConnection(tail(response, -1)),
                               stringsAsFactors=FALSE, sep="|", header=FALSE))
  names(response) <- c("as", "ip", "bgp_prefix", "cc",
                       "registry", "allocated", "as_name")
  response[response=="NA"] <- NA

  return(dplyr::tbl_df(response))

}

.bulk_peer <- function(ips, timeout=getOption("timeout")) {

  host <- "peer.whois.cymru.com"
  # host <- "v4-peer.whois.cymru.com"
  port <- 43

  # setup query
  cmd <- "begin\nverbose\n"
  ips_c <- paste(unlist(ips), collapse="\n")
  cmd <- sprintf("%s%s\nend\n", cmd, ips_c)

  # setup connection and post query
  con <- sock(host=host, port=port, blocking=TRUE, open="r+", timeout=timeout)
  if (is.null(con$result)) {
    message("Error opening connection to v4-peer.whois.cymru.com")
    return(dplyr::tbl_df(data.frame(peer_as=rep(NA, length(ips)),
                                    ip=rep(NA, length(ips)),
                                    bgp_prefix=rep(NA, length(ips)),
                                    cc=rep(NA, length(ips)),
                                    registry=rep(NA, length(ips)),
                                    allocated=rep(NA, length(ips)),
                                    peer_as_name=rep(NA, length(ips)))))
  }

  con <- con$result
  cat(cmd, file=con)
  response <- readLines(con)
  close(con)

  if (length(response) == 0) {
    message("Error reading from connection to v4-peer.whois.cymru.com")
    return(dplyr::tbl_df(data.frame(peer_as=rep(NA, length(ips)),
                                    ip=rep(NA, length(ips)),
                                    bgp_prefix=rep(NA, length(ips)),
                                    cc=rep(NA, length(ips)),
                                    registry=rep(NA, length(ips)),
                                    allocated=rep(NA, length(ips)),
                                    peer_as_name=rep(NA, length(ips)))))
  }


  # trim header, split fields and convert results
  response <- trim_df(read.csv(textConnection(tail(response, -1)),
                               stringsAsFactors=FALSE, sep="|", header=FALSE))
  names(response) <- c("peer_as", "ip", "bgp_prefix", "cc",
                       "registry", "allocated", "peer_as_name")
  response[response=="NA"] <- NA

  return(dplyr::tbl_df(response))

}

.bulk_origin_asn <- function(asns, timeout=getOption("timeout")) {

  host <- "v4.whois.cymru.com"
  port <- 43

  # setup query
  cmd <- "begin\nverbose\n"
  ips <- paste(unlist(ifelse(grepl("^AS", asns), asns,
                             sprintf("AS%s", asns))), collapse="\n")
  cmd <- sprintf("%s%s\nend\n", cmd, ips)

  # setup connection and post query
  con <- sock(host=host, port=port, blocking=TRUE, open="r+", timeout=timeout)
  if (is.null(con$result)) {
    message("Error opening connection to v4.whois.cymru.com")
    return(dplyr::tbl_df(data.frame(as=rep(NA, length(asns)),
                                    cc=rep(NA, length(asns)),
                                    registry=rep(NA, length(asns)),
                                    allocated=rep(NA, length(asns)),
                                    as_name=rep(NA, length(asns)))))
  }

  con <- con$result
  cat(cmd, file=con)
  response <- readLines(con)
  close(con)

  if (length(response) == 0) {
    message("Error reading from connection to v4.whois.cymru.com")
    return(dplyr::tbl_df(data.frame(as=rep(NA, length(asns)),
                                    cc=rep(NA, length(asns)),
                                    registry=rep(NA, length(asns)),
                                    allocated=rep(NA, length(asns)),
                                    as_name=rep(NA, length(asns)))))
  }

  # trim header, split fields and convert results
  response <- trim_df(read.csv(textConnection(tail(response, -1)),
                               stringsAsFactors=FALSE, sep="|", header=FALSE))
  names(response) <- c("as", "cc", "registry", "allocated", "as_name")
  response[response=="NA"] <- NA

  return(dplyr::tbl_df(response))

}

#' Retrieves BGP Origin ASN info for a list of IPv4 addresses
#'
#' @param ips vector of IPv4 address (character - dotted-decimal)
#' @param timeout	numeric: the timeout (in seconds) to be used for this connection.
#'        Beware that some OSes may treat very large values as zero: however the
#'        POSIX standard requires values up to 31 days to be supported.
#' @return data frame of BGP Origin ASN lookup results
#'   \itemize{
#'     \item \code{as} - AS #
#'     \item \code{ip} - IPv4 (passed in)
#'     \item \code{bgp_refix} - BGP CIDR
#'     \item \code{cc} - Country code
#'     \item \code{registry} - Registry it falls under
#'     \item \code{allocated} - date it was allocated
#'     \item \code{as_ame} - AS name
#'   }
#'   If a socket connection cannot be made (i.e. a network problem on your
#'   end or a service/network problem on their end), all columns will be
#'   \code{NA}.
#' @note The Team Cymru's service is NOT a GeoIP service! Do not use this
#'       function for that as your results will not be accurate.
#'       Data is updated every 4 hours. Also,
#'       A direct connection to TCP Port 43 (WHOIS) is required for most of these
#'       API functions to work properly.
#' @seealso \url{http://www.team-cymru.org/IP-ASN-mapping.html}
#' @export
#' @examples \dontrun{
#' bulk_origin(c("68.22.187.5", "207.229.165.18", "198.6.1.65"))
#' }
bulk_origin <- memoise::memoise(.bulk_origin)


#' Retrieves BGP Peer ASN info for a list of IPv4 addresses
#'
#' @param ips vector of IPv4 address (character - dotted-decimal)
#' @param timeout	numeric: the timeout (in seconds) to be used for this connection.
#'        Beware that some OSes may treat very large values as zero: however the
#'        POSIX standard requires values up to 31 days to be supported.
#' @return data frame of BGP Peer ASN lookup results
#'   \itemize{
#'     \item \code{peer_as} - peer AS #
#'     \item \code{ip} - IPv4 (passsed in)
#'     \item \code{bgp_prefix} - BGP CIDR block
#'     \item \code{cc} - Country code
#'     \item \code{registry} - Registry it falls under
#'     \item \code{allocated} - date allocated
#'     \item \code{peer_as_name} - peer name
#'   }
#'   If a socket connection cannot be made (i.e. a network problem on your
#'   end or a service/network problem on their end), all columns will be
#'   \code{NA}.
#' @note The Team Cymru's service is NOT a GeoIP service! Do not use this
#'       function for that as your results will not be accurate.
#'       Data is updated every 4 hours. Also,
#'       A direct connection to TCP Port 43 (WHOIS) is required for most of these
#'       API functions to work properly.
#' @seealso \url{http://www.team-cymru.org/IP-ASN-mapping.html}
#' @export
#' @examples \dontrun{
#' bulk_peer(c("68.22.187.5", "207.229.165.18", "198.6.1.65"))
#' }
bulk_peer <- memoise::memoise(.bulk_peer)


#' Retrieves BGP Origin ASN info for a list of ASN ids
#'
#' @param asns character vector of ASN ids (character)
#' @param timeout	numeric: the timeout (in seconds) to be used for this connection.
#'        Beware that some OSes may treat very large values as zero: however the
#'        POSIX standard requires values up to 31 days to be supported.
#' @return data frame of BGP Origin ASN lookup results
#'   \itemize{
#'     \item \code{as} - AS #
#'     \item \code{cc} - Country code
#'     \item \code{registry} - registry it falls under
#'     \item \code{allocated} - when it was allocated
#'     \item \code{as_name} - name associated with the allocation
#'   }
#'   If a socket connection cannot be made (i.e. a network problem on your
#'   end or a service/network problem on their end), all columns will be
#'   \code{NA}.
#' @note The Team Cymru's service is NOT a GeoIP service! Do not use this
#'       function for that as your results will not be accurate.
#'       Data is updated every 4 hours. Also,
#'       A direct connection to TCP Port 43 (WHOIS) is required for most of these
#'       API functions to work properly.
#' @seealso \url{http://www.team-cymru.org/IP-ASN-mapping.html}
#' @export
#' @examples \dontrun{
#' bulk_origin_asn(c(22822, 1273, 2381, 2603, 2914, 3257, 3356, 11164,
#'                   174, 286, 1299, 2914, 3257, 3356, 3549, 22822))
#' }
bulk_origin_asn <- memoise::memoise(.bulk_origin_asn)
